src/compute/src/render/reduce.rs

```diff
@@ -630,47 +653,106 @@ where
-            .reduce_abelian::<_, RowSpine<_, _, _, _>>("ReduceInaccumulable", move |_, _, t| {
-                t.push(((), 1))
-            })
-            .as_collection(|k, _| k.clone())
+        if validating {
+            let (oks, errs) =
+                build_reduce_inaccumulable_distinct::<_, Result<(), String>>(debug_name, partial)
+                    .as_collection(|k, v| (k.clone(), v.clone()))
+                    .map_fallible("Demux Errors", move |(key, result)| match result {
+                        Ok(()) => Ok(key),
+                        Err(m) => Err(DataflowError::from(EvalError::Internal(m))),
+                    });
+            err_output = Some(errs);
+            partial = oks;
+        } else {
+            partial = build_reduce_inaccumulable_distinct::<_, ()>(debug_name, partial)
+                .as_collection(|k, _| k.clone());
+        }
     }
 
-    let debug_name = debug_name.to_string();
-    partial
-        .arrange_named::<RowSpine<_, Row, _, _>>("Arranged ReduceInaccumulable")
-        .reduce_abelian("ReduceInaccumulable", {
+    let arranged = partial.arrange_named::<RowSpine<_, Row, _, _>>("Arranged ReduceInaccumulable");
+    let oks = arranged.reduce_abelian::<_, RowSpine<_, _, _, _>>("ReduceInaccumulable", {
         let mut row_buf = Row::default();
         move |_key, source, target| {
-            // Negative counts would be surprising, but until we are 100% certain we wont
-            // see them, we should report when we do. We may want to bake even more info
-            // in here in the future.
-            if source.iter().any(|(_val, cnt)| cnt < &0) {
-                // XXX: This reports user data, which we perhaps should not do!
-                for (val, cnt) in source.iter() {
-                    if *cnt < 0 {
-                        warn!("[customer-data] Negative accumulation in ReduceInaccumulable: {val:?} with count {cnt:?} in dataflow {debug_name}");
-                        soft_assert_or_log!(false, "Negative accumulation in ReduceInaccumulable");
+            // We respect the multiplicity here (unlike in hierarchical aggregation)
+            // because we don't know that the aggregation method is not sensitive
+            // to the number of records.
+            let iter = source.iter().flat_map(|(v, w)| {
+                // Note that in the non-positive case, this is wrong, but harmless because
+                // our other reduction will produce an error.
+                let count = usize::try_from(*w).unwrap_or(0);
+                std::iter::repeat(v.iter().next().unwrap()).take(count)
+            });
+            row_buf.packer().push(func.eval(iter, &RowArena::new()));
+            target.push((row_buf.clone(), 1));
+        }
+    });
+
+    if validating && err_output.is_none() {
+        let debug_name = debug_name.to_string();
+        let errs = arranged.reduce_abelian::<_, ErrValSpine<_, _, _>>(
```

Comment:
Yes, indeed. It might not be too bad to introduce a `conditional_reduce_pair` method in `ReduceExt`, but not sure if that would make the code here cleaner?